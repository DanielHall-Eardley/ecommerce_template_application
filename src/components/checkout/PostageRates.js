import React from 'react'
import styles from './PostageRates.module.css'

export default({rates, shipmentId, selectedRates, setRates}) => {
  
  /*This function checks if an individual postage rate 
  is selected and returns a boolean that is used to 
  conditionally add or remove a highlight class*/
  const checkIfSelected = rateId => {
    const shipmentIndex = selectedRates.findIndex(rate => rate.shipmentId === shipmentId)
    
    if (shipmentIndex === -1) {
      return false
    }

    if (selectedRates[shipmentIndex].rateId === rateId) {
      return true
    }

    return false
  }

  
  /*This function updates the selected postage rate for an individual
  product by passing an object containing the shipment id for
  the product and the id for the currently selected postage rate into local state*/
  const updatePostageRate = (event, rateId) => {
    event.preventDefault()
    
    /*check to see if a postage rate has alreay been selected*/
    const shipmentIndex = selectedRates.findIndex(rate => rate.shipmentId === shipmentId)

    const rateObj = {
      shipmentId,
      rateId
    }

    //If no rate has been selected append a new rate object to end of array
    if (shipmentIndex === -1) {
      return setRates([...selectedRates, rateObj])
    }

    //Else replace the current rate object
    const newArray = [
      ...selectedRates.slice(0, shipmentIndex),
      rateObj,
      ...selectedRates.slice(shipmentIndex + 1) 
    ]
   
    setRates(newArray)
  }
  
  const rateArray = rates.map(rate => {
    return (
      <li key={rate.rateId}
        className={
          checkIfSelected(rate.rateId) ?
          styles.selected :
          null
        }>
        <button 
          aria-label='select postage rate'
          onClick={(event) => updatePostageRate(event, rate.rateId)}>
          <span>
            { !rate.guaranteedDeliveryTime ? 
              `Estimated delivery time: ${rate.deliveryTime} days` :
              `Guaranteed delivery in: ${rate.deliveryTime} days`
            }
          </span>
          <span>{rate.serviceName}</span>
          <div className={styles.carrierAndFee}>
            <span className={styles.fee}>${rate.fee}</span>
            <span>{rate.carrier}</span>
          </div>
        </button>
      </li>
    )
  })

  return <ul className={styles.rateList}>{rateArray}</ul>
}