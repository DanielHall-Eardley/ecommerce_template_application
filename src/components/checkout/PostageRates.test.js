import React from 'react'

import {shallow} from 'enzyme'
import PostageRates from './PostageRates.js'

let props;
let wrapper
let event

beforeEach(() => {
  props = {
    rates: [
      {
        rateId: 1,
        guaranteedDeliveryTime: true,
        deliveryTime: 2,
        carrier: 'A postal service',
        fee: 10.95,
        serviceName: 'Express'
      },
      {
        rateId: 2,
        guaranteedDeliveryTime: false,
        deliveryTime: 6,
        carrier: 'A postal service',
        fee: 4.95,
        serviceName: 'Standard'
      }
    ],
    shipmentId: 1,
    selectedRates: [
      {
        shipmentId: 3,
        rateId: 4
      },
      {
        shipmentId: 1,
        rateId: 2
      },
      {
        shipmentId: 2,
        rateId: 1
      }
    ],
    setRates: jest.fn()
  }

  wrapper = shallow(<PostageRates {...props}/>)
  event = {
    preventDefault: jest.fn()
  }
})

test('snapshot matches and is up to date', () => {
  expect(wrapper).toMatchSnapshot()
})

test('test that only the selected postage rate receives a class to highlight it', () => {
  expect(wrapper.find('ul').childAt(1).hasClass('selected')).toEqual(true)
  expect(wrapper.find('ul').childAt(0).hasClass('selected')).toEqual(false)
})

test('test that the selected rates are updated correctly', () => {
  const expectedResult = [
    {
      shipmentId: 3,
      rateId: 4
    },
    {
      shipmentId: 1,
      rateId: 1
    },
    {
      shipmentId: 2,
      rateId: 1
    }
  ]

  const button = wrapper.find('ul').childAt(0).find('button')
  button.simulate('click', event)
  expect(event.preventDefault).toHaveBeenCalledTimes(1)
  expect(props.setRates).toBeCalledWith(expectedResult)
})

test('test that rates are added correctly', () => {
  const expectedResult = [
    {
      shipmentId: 3,
      rateId: 4
    },
    {
      shipmentId: 1,
      rateId: 2
    }
  ]

  props.selectedRates = [
    {
      shipmentId: 3,
      rateId: 4
    },
  ]
  wrapper = shallow(<PostageRates {...props}/>)

  const button = wrapper.find('ul').childAt(1).find('button')
  button.simulate('click', event)
  expect(event.preventDefault).toHaveBeenCalledTimes(1)
  expect(props.setRates).toBeCalledWith(expectedResult)
})