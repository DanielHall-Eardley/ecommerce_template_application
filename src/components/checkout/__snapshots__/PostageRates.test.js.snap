// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot matches and is up to date 1`] = `ShallowWrapper {}`;
